### PR TITLE
Add support for `ResourceShare` resource

### PR DESCRIPTION
**Pull request by michaelhtm**
Issue [#492](https://github.com/aws-controllers-k8s/community/issues/492)

Description of changes:
* Generated `ResourceShare`
* Added e2e tests

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: apis/v1alpha1/ack-generate-metadata.yaml [hunk 1]
```diff
@@ -1,13 +1,13 @@
 ack_generate_info:
-  build_date: "2024-09-05T21:11:27Z"
+  build_date: "2024-09-10T22:55:43Z"
   build_hash: f8f98563404066ac3340db0a049d2e530e5c51cc
   go_version: go1.23.0
-  version: v0.38.1
-api_directory_checksum: ce00c785ba2613ae96679fd158c14ba8fa8c879f
+  version: v0.38.1-dirty
```
Comment (a-hilaly) [inline apis/v1alpha1/ack-generate-metadata.yaml:5]: Can we use latest version with no modification?

File: apis/v1alpha1/generator.yaml [hunk 2]
```diff
@@ -1,5 +1,27 @@
...
-      - ResourceShare
+resources:
+  ResourceShare:
+    fields:
+      Tags:
+        from:
+          operation: TagResource
+          path: Tags
+        compare:
+          is_ignored: True
```
Comment (a-hilaly) [inline apis/v1alpha1/generator.yaml:16]: 👍

File: go.mod [hunk 3]
```diff
@@ -7,7 +7,9 @@ toolchain go1.23.0
 require (
 	github.com/aws-controllers-k8s/runtime v0.38.0
 	github.com/aws/aws-sdk-go v1.49.0
+	github.com/go-logr/logr v1.4.1
 	github.com/spf13/pflag v1.0.5
+	k8s.io/api v0.30.1
 	k8s.io/apimachinery v0.30.1
```
Comment (a-hilaly) [inline go.mod:13]: is this done by `go mod tidy` ?

File: pkg/resource/resource_share/hooks.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
+package resource_share
```
Comment (a-hilaly) [inline pkg/resource/resource_share/hooks.go:1]: nit: license header

File: pkg/resource/resource_share/hooks.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+import (
+	"context"
+
+	svcapitypes "github.com/aws-controllers-k8s/ram-controller/apis/v1alpha1"
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	svcsdk "github.com/aws/aws-sdk-go/service/ram"
```
Comment (a-hilaly) [inline pkg/resource/resource_share/hooks.go:9]: nit: ordering

File: templates/hooks/resource_share/sdk_ensure_tags.go.tpl [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+             panic("resource manager's EnsureTags method received resource with nil CR object")
+     }
+     defaultTags := ackrt.GetDefaultTags(&rm.cfg, r.ko, md)
+     var existingTags []*svcapitypes.Tag
+     existingTags = r.ko.Spec.Tags
+
+     resourceTags := ToACKTags(existingTags)
+     tags := acktags.Merge(resourceTags, defaultTags)
+     r.ko.Spec.Tags = FromACKTags(tags)
+     return nil
```
Comment (a-hilaly) [inline templates/hooks/resource_share/sdk_ensure_tags.go.tpl]: I have never seen this in another controller? what's the goal here

File: templates/hooks/resource_share/sdk_update_pre_build_request.go.tpl [hunk 7]
```diff
@@ -0,0 +1,8 @@
+	if delta.DifferentAt("Spec.Tags") {
+		if err := rm.syncTags(ctx, desired, latest); err != nil {
+			return nil, err
+		}
+	}
+	if !delta.DifferentExcept("Spec.Tags") {
+		return desired, nil
+	}
```
Comment (a-hilaly) [inline templates/hooks/resource_share/sdk_update_pre_build_request.go.tpl:8]: sweet!

**Review by a-hilaly — COMMENTED** [review]
Excellent work on this @michaelhtm !! :)
I left a few comments and questions inline

File: go.mod (hunk 3 above)
Comment (michaelhtm) [reply go.mod:13]: yes

**Review by a-hilaly — APPROVED** [review]
/lgtm

**Pull request merged**